File: plugins/premiere-uxp/src/apply/sfxMusic.ts
/**
 * Apply accepted `sfx`/`music` items as clips on dedicated audio tracks (D09 brief §Scope 1):
 * downloaded owned assets only (the licence predicate is re-checked client-side inside
 * `@montaj/shared-apply`'s `buildApplyPlan` — a partner asset never reaches this module as an
 * `AudioClipOp`, only as an `AudioClipRefusal` the caller surfaces as the D43 "cloud render
 * only" badge), gain and fades as clip audio keyframes, ducking approximated as gain keyframes
 * on the dialogue track.
 *
 * This module never re-derives the plan itself — `@montaj/shared-apply`'s `buildApplyPlan` is
 * the single source of truth both plugins call, so a change to the licence gate or the fade/duck
 * shape only ever needs to change in one place (see that package's `planBuilder.ts` doc comment
 * for why a Python port of the *logic* was rejected in favour of one shared TS builder).
 */
import { buildApplyPlan } from "@montaj/shared-apply";
import type { ApplyPlanItem, AudioClipOp, AudioClipRefusal } from "@montaj/shared-apply";

import type { GainKeyframeInput, PremiereHost } from "../host/premiere.js";

/** Gain floor used for a fade's silent endpoint — Premiere's audio gain keyframes have no true
 * "-infinity" value in the mock/host surface this WP targets, so -60 dB stands in for silence. */
const SILENT_GAIN_DB = -60;

const msToFrames = (ms: number, fps: number): number => Math.round((ms / 1000) * fps);

export interface ApplySfxMusicOptions {
  readonly projectId: string;
  readonly items: readonly ApplyPlanItem[];
  /** Local file path for each accepted item's downloaded asset (already fetched by the caller —
   * this module never performs the download itself, matching every other C06 apply mode, which
   * takes a local path/handle rather than a URL). Missing entries are skipped, not guessed. */
  readonly assetLocalPaths: ReadonlyMap<string, string>;
  readonly fps: number;
  /** The dialogue track item to approximate ducking against, if the sequence has one mapped. */
  readonly dialogueTrackItemId?: string;
}

export interface PlacedAudioClip {
  readonly itemId: string;
  readonly trackItemId: string;
  readonly kind: AudioClipOp["kind"];
}

export interface ApplySfxMusicResult {
  readonly placed: readonly PlacedAudioClip[];
  /** Accepted items with no local asset path yet supplied — not an error, just "not ready". */
  readonly skipped: readonly { itemId: string; reason: "asset-not-downloaded" }[];
  readonly refusals: readonly AudioClipRefusal[];
}

export function buildGainKeyframes(op: AudioClipOp, fps: number): GainKeyframeInput[] {
  const startFrames = msToFrames(op.startMs, fps);
  const endFrames = msToFrames(op.startMs + op.durationMs, fps);
  const keyframes: GainKeyframeInput[] = [];

  if (op.fade.fadeInMs > 0) {
    keyframes.push({ atFrames: startFrames, gainDb: SILENT_GAIN_DB });
    keyframes.push({
      atFrames: startFrames + msToFrames(op.fade.fadeInMs, fps),
      gainDb: op.gainDb,
    });
  } else {
    keyframes.push({ atFrames: startFrames, gainDb: op.gainDb });
  }

  if (op.fade.fadeOutMs > 0) {
    keyframes.push({
      atFrames: Math.max(endFrames - msToFrames(op.fade.fadeOutMs, fps), startFrames),
      gainDb: op.gainDb,
    });
    keyframes.push({ atFrames: endFrames, gainDb: SILENT_GAIN_DB });
  }

  return keyframes;
}

/** Ducking approximation (brief §Scope 1): the dialogue track dips by `duck.depthDb` over the
 * sfx/music clip's own range, ramping in over `attackMs` and back out over `releaseMs`. This is
 * a coarse stand-in for a real sidechain compressor — documented as an approximation, not a
 * faithful ducking model, matching the Resolve side's own documented `DynamicZoomEase`-style
 * limitation for zooms. */
export function buildDuckKeyframes(op: AudioClipOp, fps: number): GainKeyframeInput[] | undefined {
  if (!op.duck) return undefined;
  const { depthDb, attackMs, releaseMs } = op.duck;
  const startFrames = msToFrames(op.startMs, fps);
  const endFrames = msToFrames(op.startMs + op.durationMs, fps);
  const attackEndFrames = Math.min(startFrames + msToFrames(attackMs, fps), endFrames);
  const releaseStartFrames = Math.max(endFrames - msToFrames(releaseMs, fps), attackEndFrames);
  return [
    { atFrames: startFrames, gainDb: 0 },
    { atFrames: attackEndFrames, gainDb: depthDb },
    { atFrames: releaseStartFrames, gainDb: depthDb },
    { atFrames: endFrames, gainDb: 0 },
  ];
}

export async function applyAcceptedSfxMusic(
  host: PremiereHost,
  options: ApplySfxMusicOptions,
): Promise<ApplySfxMusicResult> {
  const plan = buildApplyPlan(options.items);
  const audioOps = plan.ops.filter((op): op is AudioClipOp => op.op === "audioClip");

  const placed: PlacedAudioClip[] = [];
  const skipped: { itemId: string; reason: "asset-not-downloaded" }[] = [];

  for (const op of audioOps) {
    const localPath = options.assetLocalPaths.get(op.assetId);
    if (!localPath) {
      skipped.push({ itemId: op.itemId, reason: "asset-not-downloaded" });
      continue;
    }

    const track = await host.ensureTrack({ kind: "audio", name: op.trackName });
    const bin = await host.importMediaToBin({ sourcePath: localPath });
    const { trackItemId } = await host.placeOnTrack({
      itemId: bin.itemId,
      trackIndex: track.trackIndex,
      startFrames: msToFrames(op.startMs, options.fps),
      durationFrames: msToFrames(op.durationMs, options.fps),
    });

    await host.setClipGainKeyframes(trackItemId, buildGainKeyframes(op, options.fps));

    const duckKeyframes = buildDuckKeyframes(op, options.fps);
    if (duckKeyframes && options.dialogueTrackItemId) {
      await host.setClipGainKeyframes(options.dialogueTrackItemId, duckKeyframes);
    }

    await host.setItemMetadata(trackItemId, {
      aksharo: { projectId: options.projectId, itemId: op.itemId, rev: 0 },
    });

    placed.push({ itemId: op.itemId, trackItemId, kind: op.kind });
  }

  return { placed, skipped, refusals: plan.audioRefusals };
}

File: plugins/resolve/aksharo_core_app/sfx_music.py
"""Apply accepted `sfx`/`music` pass items to a Resolve timeline (D09 brief §Scope 2).

`05-system-architecture.md`/CONTRACTS §2 amendment: audio clips on new tracks
(`MediaPool.ImportMedia` + `AppendToTimeline` on an audio track — `ResolveHost.import_audio_clip`),
fades via clip properties, ducking as keyframed volume where the API allows (documented limit,
see `host/resolve.py`'s `set_volume_keyframes` docstring). Only owned, panel-licensed assets are
ever placed — `licence.py`'s `check_licence_for_panel` is re-checked here, same rule as
`plugins/premiere-uxp/src/apply/sfxMusic.ts`, D43's "cloud render only" gate.
"""

from __future__ import annotations

from dataclasses import dataclass

from aksharo_core_app.host.resolve import ResolveHost, TimelineHandle, TimelineItemHandle
from aksharo_core_app.licence import (
    CLOUD_RENDER_ONLY_MESSAGE,
    LicenceSnapshot,
    check_licence_for_panel,
)
from aksharo_core_app.markers import HostMapping, stamp_item

SFX_TRACK_NAME = "Aksharo SFX"
MUSIC_TRACK_NAME = "Aksharo Music"
SILENT_VOLUME_DB = -60.0


@dataclass(frozen=True, slots=True)
class DuckSpec:
    depth_db: float
    attack_ms: float
    release_ms: float


@dataclass(frozen=True, slots=True)
class SfxMusicItem:
    item_id: str
    project_id: str
    kind: str  # "sfx" | "music"
    asset_id: str
    start_ms: float
    duration_ms: float
    gain_db: float
    fade_in_ms: float
    fade_out_ms: float
    duck: DuckSpec | None
    licence_snapshot: LicenceSnapshot
    rev: int = 0


@dataclass(frozen=True, slots=True)
class PlacedAudioClip:
    item_id: str
    track_item_id: str
    kind: str


@dataclass(frozen=True, slots=True)
class AudioClipRefusal:
    item_id: str
    kind: str
    asset_id: str
    reasons: tuple[str, ...]
    message: str = CLOUD_RENDER_ONLY_MESSAGE


@dataclass(frozen=True, slots=True)
class ApplySfxMusicResult:
    placed: tuple[PlacedAudioClip, ...]
    skipped: tuple[str, ...]  # item ids with no local asset path supplied
    refusals: tuple[AudioClipRefusal, ...]


def ms_to_frame(ms: float, fps: float) -> int:
    return round(ms / 1000.0 * fps)


def build_volume_keyframes(
    item: SfxMusicItem, fps: float
) -> list[tuple[int, float]]:
    """Fade-in/out volume keyframes (brief: "fades via clip properties")."""
    start_frame = ms_to_frame(item.start_ms, fps)
    end_frame = ms_to_frame(item.start_ms + item.duration_ms, fps)
    keyframes: list[tuple[int, float]] = []

    if item.fade_in_ms > 0:
        keyframes.append((start_frame, SILENT_VOLUME_DB))
        keyframes.append((start_frame + ms_to_frame(item.fade_in_ms, fps), item.gain_db))
    else:
        keyframes.append((start_frame, item.gain_db))

    if item.fade_out_ms > 0:
        keyframes.append(
            (max(end_frame - ms_to_frame(item.fade_out_ms, fps), start_frame), item.gain_db)
        )
        keyframes.append((end_frame, SILENT_VOLUME_DB))

    return keyframes


def build_duck_keyframes(item: SfxMusicItem, fps: float) -> list[tuple[int, float]] | None:
    """Ducking approximation on the dialogue track (documented approximation, matching
    `plugins/premiere-uxp/src/apply/sfxMusic.ts`'s own `buildDuckKeyframes`)."""
    if item.duck is None:
        return None
    start_frame = ms_to_frame(item.start_ms, fps)
    end_frame = ms_to_frame(item.start_ms + item.duration_ms, fps)
    attack_end = min(start_frame + ms_to_frame(item.duck.attack_ms, fps), end_frame)
    release_start = max(end_frame - ms_to_frame(item.duck.release_ms, fps), attack_end)
    return [
        (start_frame, 0.0),
        (attack_end, item.duck.depth_db),
        (release_start, item.duck.depth_db),
        (end_frame, 0.0),
    ]


def apply_sfx_music(
    host: ResolveHost,
    timeline: TimelineHandle,
    items: list[SfxMusicItem],
    asset_local_paths: dict[str, str],
    fps: float,
    dialogue_track_item: TimelineItemHandle | None = None,
) -> ApplySfxMusicResult:
    placed: list[PlacedAudioClip] = []
    skipped: list[str] = []
    refusals: list[AudioClipRefusal] = []
    # `ResolveHost.add_track` (per its Protocol docstring, "AddTrack(trackType) then name it via
    # SetTrackName") documents no idempotent "find by name" behaviour the way
    # `PremiereHost.ensureTrack` does — this dict is this function's own within-call dedupe so
    # two sfx items in the same `apply_sfx_music` call share one "Aksharo SFX" track rather than
    # each getting their own. It does NOT persist across separate calls/re-applies: a second
    # `apply_sfx_music` run on an already-built timeline will still add a new track, which is
    # flagged as an open item for a follow-up WP (mirroring the marker-based re-sync this module
    # does not yet perform for its own tracks, only for its own clips via `stamp_item`).
    track_index_by_name: dict[str, int] = {}

    for item in items:
        check = check_licence_for_panel(item.licence_snapshot)
        if not check.allowed:
            refusals.append(
                AudioClipRefusal(
                    item_id=item.item_id, kind=item.kind, asset_id=item.asset_id,
                    reasons=check.reasons,
                )
            )
            continue

        local_path = asset_local_paths.get(item.asset_id)
        if local_path is None:
            skipped.append(item.item_id)
            continue

        track_name = SFX_TRACK_NAME if item.kind == "sfx" else MUSIC_TRACK_NAME
        track_index = track_index_by_name.get(track_name)
        if track_index is None:
            track_index = host.add_track(timeline, "audio", track_name)
            track_index_by_name[track_name] = track_index
        start_frame = ms_to_frame(item.start_ms, fps)
        end_frame = ms_to_frame(item.start_ms + item.duration_ms, fps)
        clip = host.import_audio_clip(timeline, track_index, start_frame, end_frame, local_path)

        host.set_volume_keyframes(clip, build_volume_keyframes(item, fps))
        duck_keyframes = build_duck_keyframes(item, fps)
        if duck_keyframes is not None and dialogue_track_item is not None:
            host.set_volume_keyframes(dialogue_track_item, duck_keyframes)

        mapping = HostMapping(
            project_id=item.project_id, ref_kind="itemId", ref_id=item.item_id, rev=item.rev
        )
        stamp_item(host, clip, mapping)
        placed.append(
            PlacedAudioClip(item_id=item.item_id, track_item_id=clip.item_id, kind=item.kind)
        )

    return ApplySfxMusicResult(
        placed=tuple(placed), skipped=tuple(skipped), refusals=tuple(refusals)
    )
